Add an initial formal specification

javierdiaz72 opened the pull request:
This PR adds an initial version of a formal specification. This version includes the relevant material from the [Shelley](https://github.com/intersectmbo/cardano-ledger/releases/latest/download/shelley-ledger.pdf) and [Babbage](https://github.com/intersectmbo/cardano-ledger/releases/latest/download/babbage-ledger.pdf) ledger specs, plus the following fixes:

1. There is an oversight in the `Decentralized` rule in Fig. 68 of the Shelley ledger spec, namely that no environment was passed in the call to the OCERT transition system.
2. The OCERT rule has a new predicate failure: We newly require (in the context of Fig. 66 in the Shelley ledger spec) the condition _n_ ≤ _m_ + 1, a violation of which corresponds to a _CounterOverIncremented_ failure.
3. There is an error in Fig. 69 of the Shelley ledger spec: Instead of `ℙ PrtclEnv`, it should be `PrtclEnv`.
4. There is an error in Sec. 12.13 of the Shelley ledger spec: Instead of "It calls BHEAD, PRTCL, and BBODY as sub-rules." it should be "It calls TICK, TICKN, PRTCL, and BBODY as sub-rules."; and similarly in the introduction paragraph of Sec. 12.
5. The former _VRFKeyBadLeaderValue_ and _VRFKeyBadNonce_ failures are replaced by the new single _VRFKeyBadProof_ failure.

javierdiaz72 commented:
Note: No integration with Nix is added yet so, in order to build the PDF file, you need to:

1. Install the `latexmk` package.
2. Run `make`.

jorisdral commented:
One nice follow-up for this work would be to have GHA render and upload this doc to the github-pages website

jmchapman commented on docs/formal-spec/chain.tex:
```diff
@@ -0,0 +1,1061 @@
+ \label{fig:rules:tick-nonce}
+\end{figure}
+
+\subsection{Update Nonce Transition}
```
Has this change been implemented in the consensus code already? If so I think we should add it here as we we want to be in sync with the implementation and also up-to-date with Conway.

jmchapman replied:
On the other hand we could do babbage only first and then do this separately if it's non-trivial.

amesgen replied:
> Has this change been implemented in the consensus code already?

Yes, it has been implemented in #927.

javierdiaz72 replied:
Thanks @amesgen for confirming.

@amesgen , @jmchapman : I think we should gather the changes from Conway that have already been implemented in the code and add them to our spec here. What do you think?

javierdiaz72 replied:
Resolved in b18f6df47e87f04b66e90cab890bca6c3e306100, please check.

amesgen replied:
There should be no changes to `docs/formal-spec/properties.tex`, these properties should refer to $\mathsf{StabilityWindow}$. Apart from that, it looks good, can you maybe also add some (foot)note somewhere that we used $\mathsf{StabilityWindow}$ for UPDN in the past (pre-Conway)?

javierdiaz72 replied:
> There should be no changes to `docs/formal-spec/properties.tex`, these properties should refer to $\mathsf{StabilityWindow}$. 

Fixed in de32ec24065fe096b7b3f13d9a19dc7b262589ab, please check.

> can you maybe also add some (foot)note somewhere that we used $\mathsf{StabilityWindow}$ for UPDN in the past (pre-Conway)?

Fixed in 6afee9445e12c9aea6d579755b3a67d3fa52291f, please check.

amesgen replied:
Thanks, looks good :+1:

jmchapman reviewed (approved):
This is a great start to the project and is now up to date with Conway.

amesgen commented:
Rendered pdf as of 6afee9445e12c9aea6d579755b3a67d3fa52291f for convenience: [consensus-spec.pdf](https://github.com/IntersectMBO/ouroboros-consensus/files/14573199/consensus-spec.pdf)

javierdiaz72 commented:
FTR: The following is the list of changes to be included in the next iteration of the spec, according to what was discussed during the mob review:

1. Remove all references to block bodies
2. Remove the BBODY transition system (and, consequently, remove the LEDGERS transition system from the Common Interface section)
4. Replace the CHAIN transition system with the CHAINHEAD transition system (and, consequently, add the NEWEPOCH transition system to the Common Interface section)
5. Keep the Properties section but add a reference to the CHAIN transition system in the Ledger spec.

javierdiaz72 commented:
FTR: There is a mistake in Fig. 75 of the Shelley spec: It should say `(e2, _, bcur, es, _, pd)` instead of `(e2, _, bcur, es, _, _, pd)`. I'll include an appropriate fix in the upcoming iteration of the spec.

javierdiaz72 commented:
Please find a new version of the spec ready for review at 7e885c501644710bd6a0a11e1c791c396828e2ae. This version includes the changes mentioned in the last 3 comments.

jasagredo commented:
I'm going to dump here some questions that I got when reading the PDF. After I'm done with the read, I will put these comments if applicable in the exact source code lines they refer to.

### Section 1.1: New epoch transition

- I would add the name of the rule to the header of the section.
- From the point of view of the Consensus layer, are the details of the transition that happens in the new epoch transition irrelevant or why is only the type shown?
- Why is this one defined before presenting the diagram of rule dependencies, or rather why is this one here and the other rules after the diagram?

### Section 2: Transition rule dependencies

- This is very different from the diagram in the shelley specs, why? In particular `TICKF` is not shown in the shelley diagram.
- There are no recursive rules in this diagram, shouldn't we remove that explanation?

### Section 3.1: Block definitions

- I find it confusing that sometimes `b` means `block`, sometimes `block body`. Same with `vrf`, sometimes it appears sometimes it doesn't. I think we should either expand the prefix or make it unique. In particular I propose:
  - Block Header Body 
    - `prev` -> `prevHeader`
    - `vk` -> `issuerVk`
    - `blockno` -> `blockNo`
    - `prf` -> `vrfPrf`
    - `bsize` -> `bodySize`
    - `bhash` -> `bodyHash`
  - Abstract functions
    - `bhHash` -> `headerHash`
    - `bHeaderSize` -> `headerSize`
  - Accessor functions
    - `bheader` -> `blockHeader` 
    - `hsig` -> `headerSig`
    - `bhbody` -> `headerBody`
    - I would list first `blockHeader`, `headerSig` and `headerBody`
    - As we are ignoring the block body, the prefix `body` is not confusing (as it could be between block body and block header body), so I also suggest prefixing the rest of functions with `body`, so `bslot` -> `bodySlot`, `bVrfRes` -> `bodyVrfRes`, etc.
    - Also all these functions can be composed with `headerBody` to give types ${\tt BHeader} \rightarrow {\tt X}$ so it would also be plausible to say `headerSlot` instead of `bodySlot`.
    - TODO: present a screenshot of what I mean

### Section 3.2: Tick Nonce Transition

- I see that `nonce` and `seed` is used interchangeably, why is this? Couldn't we set with one?
- Reference to figure is broken. Says Fig 3.2 but does not exist.
- I would add the rule name in the header of the section for clarity.

### Section 3.3: Update Nonce Transition

- I would add the rule name in the header.
- nonce candidate -> candidate nonce
- I would not put the explanation for the entropy in parenthesis. I find it relevant.

### Section 3.4: Operational certificate transition

- I would add the rule name in the header
- Could annotate the predicate failures with numbers, for later correlating with the list of failures
$$c_0 \overset{(1)}{\leq} {\tt kesPeriod~} s$$

### Section 3.5: Verifiable Random Function

- VRF helper function*s* in the figure.
- `hashKey` is used without it being defined anywhere.
- Where is `verifyVrf_seed` defined?

### Section 3.6: Protocol transition

- Add rule name in header
- Typo: canditate -> candidate
- slot is undefined, I imagine there should be a $slot := {\tt bslot~} bhb$
- Failures could be numbered in the transition diagram.
- I wonder if the failures described here are not actually failures of the VRF function instead?
- What happens if failures happen in the inner OCERT rule? don't failures bubble up?

### Section 3.7: Tick forecast transition

- Add rule name in header
- Here we destructure the NewEpochState, but this product type was never defined. Same with EpochState, ds which I imagine is DelegationState or some such, the type of fGenDelegs and genDelegs.
- This might be just my lack of knowledge, but what does $\overset{\cup}{\rightarrow}$ mean?

### Section 3.8 Chain Head Transition

- Add rule name in header
- Could failures be numbered also here?
- Figure 16: `acnt` and `ls` are unused, aren't they?

jasagredo commented:
Comments on properties:

- 4.1: Isn't it clearer to say that as $s \rightarrow s'$, then $\tilde{s} \rightarrow \tilde{s}'$ instead of $s''$?
- 4.2: in the explanation you say `initial state s` but then `nes` appears, probably just missing a `s = (nes, \tilde{s})` or some such? It is known by the context, but still.
- 4.2: You mention the BBODY rule, which I guess is also defined in SL-D5?
- I did not quite understand property 4.2.

javierdiaz72 commented:
During today's Consensus Office Hours meeting, we discussed about the interface between the consensus spec and the ledger spec. Here's the big-picture structure we've come up with (thanks @WhatisRT):

```
Interface
- Block
- BBodyEnv -> PParams (in Conway, have TriggerHF action)
- BBodyState (LState abstract?)
- NewEpochState (possibly abstract?)
- Slot


                  (BBODY, TICK, ...)
                /---- Ledger <------\
Interface Lib <-                     -- Node (CHAIN)
                \---- Consensus <---/
               (TICKF, CHAINHEAD, ...)
```

javierdiaz72 commented:
FTR, the following change will be included in the next iteration of the spec: The removal of the $extraEntropy$ in the $\mathsf{TICKN}$ rules as dictated by the Babbage spec makes the inclusion of $pp$ in the environment useless, so it should be removed.

javierdiaz72 commented:
Hi, @jasagredo! Please find below my responses to your review comments:

> ### Section 1.1: New epoch transition
> 
>     * I would add the name of the rule to the header of the section.
Fixed in 940560c8a514f6ba2e6277863fd944a2c17d42ad.
> 
>     * From the point of view of the Consensus layer, are the details of the transition that happens in the new epoch transition irrelevant or why is only the type shown?
IMO, the details should be irrelevant to the Consensus layer since those are regarded as implementation details.
> 
>     * Why is this one defined before presenting the diagram of rule dependencies, or rather why is this one here and the other rules after the diagram?
Fixed in 940560c8a514f6ba2e6277863fd944a2c17d42ad.
> 
> 
> ### Section 2: Transition rule dependencies
> 
>     * This is very different from the diagram in the shelley specs, why? In particular `TICKF` is not shown in the shelley diagram.
This diagram includes only the transition systems (and their dependencies) related to the Consensus layer.
> 
>     * There are no recursive rules in this diagram, shouldn't we remove that explanation?
Fixed in baa85e5813f089396491d3aa2c233ce6fa750b97.
> ### Section 3.1: Block definitions
> 
>     * I find it confusing that sometimes `b` means `block`, sometimes `block body`. Same with `vrf`, sometimes it appears sometimes it doesn't. I think we should either expand the prefix or make it unique. In particular I propose:
>       
>       * Block Header Body
>         
>         * `prev` -> `prevHeader`
>         * `vk` -> `issuerVk`
>         * `blockno` -> `blockNo`
>         * `prf` -> `vrfPrf`
>         * `bsize` -> `bodySize`
>         * `bhash` -> `bodyHash`
>       * Abstract functions
>         
>         * `bhHash` -> `headerHash`
>         * `bHeaderSize` -> `headerSize`
>       * Accessor functions
>         
>         * `bheader` -> `blockHeader`
>         * `hsig` -> `headerSig`
>         * `bhbody` -> `headerBody`
Fixed in 659cd85013d6be52f31951120c54472bf2c569ec.
>         * I would list first `blockHeader`, `headerSig` and `headerBody`
Actually they are already listed in that order: The list is read from left to right.
>         * As we are ignoring the block body, the prefix `body` is not confusing (as it could be between block body and block header body), so I also suggest prefixing the rest of functions with `body`, so `bslot` -> `bodySlot`, `bVrfRes` -> `bodyVrfRes`, etc.
I agree about the inconsistency in the use of the prefixes. However, the problem I see with the prefix $body$ is that it's confusing when used for the fields $bodySize$ and $bodyHash$ (e.g., $bodyBodySize$ looks weird). Therefore, I prefer to choose your second option, i.e., use a unique prefix.
>         * Also all these functions can be composed with `headerBody` to give types BHeader→X so it would also be plausible to say `headerSlot` instead of `bodySlot`.
That's true, however these functions are mostly used when only header bodies are available.
>         * TODO: present a screenshot of what I mean
> 
> 
> ### Section 3.2: Tick Nonce Transition
> 
>     * I see that `nonce` and `seed` is used interchangeably, why is this? Couldn't we set with one?
> 
>     * Reference to figure is broken. Says Fig 3.2 but does not exist.
>
>     * I would add the rule name in the header of the section for clarity.
Fixed in 7bd1fdf51affb38f2728681fdf2816614df63b46.
> 
> 
> ### Section 3.3: Update Nonce Transition
> 
>     * I would add the rule name in the header.
> 
>     * nonce candidate -> candidate nonce
> 
>     * I would not put the explanation for the entropy in parenthesis. I find it relevant.
Fixed in b9f53081074b943d9151fdd718b1c8aa4d35bdb2.
> 
> 
> ### Section 3.4: Operational certificate transition
> 
>     * I would add the rule name in the header
> 
>     * Could annotate the predicate failures with numbers, for later correlating with the list of failures
>       c0≤(1)kesPeriod s
Fixed in f4b8ec5e96082c88935770a84e173b7d4d65de84.
> 
> 
> ### Section 3.5: Verifiable Random Function
> 
>     * VRF helper function_s_ in the figure.
Fixed in 2d69d91d4371601cf145b6f56f0f0f71d9e10129.
> 
>     * `hashKey` is used without it being defined anywhere.
> 
>     * Where is `verifyVrf_seed` defined?
These are currently defined in the ledger spec but should be included in the new "base" (or "library") spec, see https://github.com/IntersectMBO/ouroboros-consensus/pull/982#issuecomment-2077689225.
> 
> 
> ### Section 3.6: Protocol transition
> 
>     * Add rule name in header
> 
>     * Typo: canditate -> candidate
> 
>     * slot is undefined, I imagine there should be a slot:=bslot bhb
> 
>     * Failures could be numbered in the transition diagram.
> 
>     * I wonder if the failures described here are not actually failures of the VRF function instead?
> 
>     * What happens if failures happen in the inner OCERT rule? don't failures bubble up?
> 
Fixed in 84510796dcda3d95b9c85fcc0b0c8f7d16a662a9.
> 
> ### Section 3.7: Tick forecast transition
> 
>     * Add rule name in header
Fixed in 19468102df496977ee943787c1196543f1a9e5d8.
> 
>     * Here we destructure the NewEpochState, but this product type was never defined. Same with EpochState, ds which I imagine is DelegationState or some such, the type of fGenDelegs and genDelegs.
These are currently defined in the ledger spec but should be included in the new "base" (or "library") spec, see https://github.com/IntersectMBO/ouroboros-consensus/pull/982#issuecomment-2077689225.
> 
>     * This might be just my lack of knowledge, but what does →∪ mean?
It's the _union override right_ operator on maps, that is, union on maps with a bias towards the right operand in case of key clashes. This is currently defined in the ledger spec.
> 
> 
> ### Section 3.8 Chain Head Transition
> 
>     * Add rule name in header
> 
>     * Could failures be numbered also here?
> 
>     * Figure 16: `acnt` and `ls` are unused, aren't they?
Fixed in bc6f767b4851149f3b8bd996c0a8eee695ba16c8.

javierdiaz72 commented:
> FTR, the following change will be included in the next iteration of the spec: The removal of the extraEntropy in the TICKN rules as dictated by the Babbage spec makes the inclusion of pp in the environment useless, so it should be removed.

Fixed in 7c07c27880cfc4eda58b21a7a932316ce9d55562.

javierdiaz72 commented:
Hi, team! We now have an improved version of the spec which is quite close to the code. Could you please review it? Thanks!

javierdiaz72 commented:
Note: f9b03fa9e62968d4a278e6b48bb5763f0c2d7602 fixes a mistake found originally in the Babbage spec (see https://github.com/IntersectMBO/cardano-ledger/pull/4287#issuecomment-2098975761).

javierdiaz72 commented:
> I think all my concerns above were resolved and/or answered. Thanks Javier!

That's great to hear, thank you, @jasagredo ! Could you perhaps give your approval to this PR?

WhatisRT reviewed (approved):
I'm not super familiar with all of these details, so I only have a minor comment.

javierdiaz72 commented on docs/formal-spec/chain.tex:
```diff
@@ -0,0 +1,976 @@
+      &
+      m \leq n \; (\ref{itm:ocert-failures-3})
+      &
+      n \leq m + 1 \; (\ref{itm:ocert-failures-4})
```
> Is that check new? It's not in the Shelley spec.
Yes, it's a new check (see the description of this PR for more info).
> 
> Also, I think I'd prefer something like n∈m,m+1. This makes it very clear that there are exactly two allowed values. The predicate failures could still be generated using these conditions.
I certainly agree, will fix this now.

WhatisRT replied:
Thanks! I only had a very quick scroll through everything, so I missed that.

javierdiaz72 replied:
Fixed in 8d4589382b38999a56a82b25fd5247076ef75c50, please check.

javierdiaz72 commented:
Hi, Team! Are you satisfied with the spec of is there still anything preventing this PR to be merged?

amesgen commented:
Ah, the only thing was that the commits aren't signed (which is required). I rebased and squashed, should now enter the merge queue once CI passed :+1:
